code/modules/power/supermatter/supermatter.dm

```diff
@@ -1,438 +1,876 @@
+		if((P.anchored || P.move_resist >= MOVE_FORCE_EXTREMELY_STRONG)) //move resist memes.
+			if(istype(P, /obj/structure/closet))
+				var/obj/structure/closet/toggle = P
+				toggle.open()
+			continue
+		if(ismob(P))
+			var/mob/M = P
+			if(M.mob_negates_gravity())
+				continue //You can't pull someone nailed to the deck
+		step_towards(P,center)
+
+/obj/machinery/power/supermatter_crystal/proc/supermatter_anomaly_gen(turf/anomalycenter, type = FLUX_ANOMALY, anomalyrange = 5)
+	var/turf/L = pick(orange(anomalyrange, anomalycenter))
+	if(L)
+		switch(type)
+			if(FLUX_ANOMALY)
+				var/obj/effect/anomaly/flux/A = new(L, 300, FALSE)
+				A.explosive = FALSE
+			if(GRAVITATIONAL_ANOMALY)
+				new /obj/effect/anomaly/grav(L, 250, FALSE)
+			if(PYRO_ANOMALY)
+				new /obj/effect/anomaly/pyro(L, 200, FALSE)
+
+
+
+/obj/machinery/power/supermatter_crystal/proc/supermatter_zap(atom/zapstart, range = 3, power)
+	if(QDELETED(zapstart))
+		return
+	. = zapstart.dir
+	if(power < 1000)
+		return
 
-	if((get_integrity() < 100) || (air.temperature > CRITICAL_TEMPERATURE))
-		return SUPERMATTER_WARNING
+	var/target_atom
+	var/mob/living/target_mob
+	var/obj/machinery/target_machine
+	var/obj/structure/target_structure
+	var/list/arctargetsmob = list()
+	var/list/arctargetsmachine = list()
+	var/list/arctargetsstructure = list()
+
+	if(prob(20)) //let's not hit all the engineers with every beam and/or segment of the arc
+		for(var/mob/living/Z in oview(zapstart, range+2))
+			arctargetsmob += Z
+	if(arctargetsmob.len)
+		var/mob/living/H = pick(arctargetsmob)
+		var/atom/A = H
+		target_mob = H
+		target_atom = A
 
-	if(air.temperature > (CRITICAL_TEMPERATURE * 0.8))
-		return SUPERMATTER_NOTIFY
+	else
+		for(var/obj/machinery/X in oview(zapstart, range + 2))
+			arctargetsmachine += X
+		if(arctargetsmachine.len)
+			var/obj/machinery/M = pick(arctargetsmachine)
+			var/atom/A = M
+			target_machine = M
+			target_atom = A
 
-	if(power > 5)
-		return SUPERMATTER_NORMAL
-	return SUPERMATTER_INACTIVE
+		else
+			for(var/obj/structure/Y in oview(zapstart, range + 2))
+				arctargetsstructure += Y
+			if(arctargetsstructure.len)
+				var/obj/structure/O = pick(arctargetsstructure)
+				var/atom/A = O
+				target_structure = O
+				target_atom = A
+
+	if(target_atom)
+		switch(power)
+			if(POWER_PENALTY_THRESHOLD to SEVERE_POWER_PENALTY_THRESHOLD)
+				zap_icon = DEFAULT_ZAP_ICON_STATE
+			if(SEVERE_POWER_PENALTY_THRESHOLD to CRITICAL_POWER_PENALTY_THRESHOLD)
+				zap_icon = SLIGHTLY_CHARGED_ZAP_ICON_STATE
+			if(CRITICAL_POWER_PENALTY_THRESHOLD to INFINITY)
+				zap_icon = OVER_9000_ZAP_ICON_STATE
+		zapstart.Beam(target_atom, icon_state = zap_icon, time = 5)
+		var/zapdir = get_dir(zapstart, target_atom)
+		if(zapdir)
+			. = zapdir
+
+	if(target_mob)
+		target_mob.electrocute_act(rand(5, 10), "Supermatter Discharge Bolt", 1, stun = FALSE)
+		if(prob(15))
+			supermatter_zap(target_mob, 5, power / 2)
+			supermatter_zap(target_mob, 5, power / 2)
+		else
```

Comment:
Duplicated lines?

Reply:
None of these are duplicates, but I went with a different style to make it more clear this is intentional.